docs/proposals/scalable-alertmanager.md

```diff
@@ -0,0 +1,78 @@
+Alertmanager is a stateful service; it stores the notification state and configured silences. By default, Alertmanager persists its state to disk every 15mins.  In the horizontally scalable Alertmanager service, we need to move this state around as the number of replicas grows and shrinks.  We also need to persist this state across restarts and rolling upgrades.
+
+**We propose** making each Alertmanager replica flush the state to object storage, under its own key that’s a combination of tenant ID + replica periodically. This state on durable storage will only be used when cold-starting the cluster.
+
+This mechanism covers multiple challenges (scaling up & down, rolling restarts, total cluster failure). **To top this off, in the implementation, we’ll always try to request state from other replicas in the cluster before ever trying to go to object storage.**
+
+### Replication & Consistency
+
+Upstream Alertmanager replicates notification state between replicas to ensure notifications are not sent more than once. Cortex Alertmanager does the same.  When we move to a model where AM instances for a given tenant only live on a subset of replicas, we have to decide how we will keep these replicas in sync.
+
+**We have an option of doing nothing** and use the existing gossip mechanism and gossip all tenants state to all replicas. The AM Router will then drop state updates for tenants which don’t shard to a given replica. I think this will be easy to implement as it requires  few changes but probably won’t scale.
+
+**However, I propose we Synchronized state over gRPC.** The upstream Alertmanager notification dispatcher uses a timeout-based approach for its notifications. Using the “peer order”, it’ll wait a certain amount of time before letting other replicas know if this notification succeeded or not. If it didn’t, the next replica in line will try to send the notification. We propose to communicate this via gRPC calls.
+
+While I propose the use of gRPC, is good to note that doing nothing is still a solid candidate for consideration but not having to maintain and operate two separate replication patterns feels like a win.
+
+### Architecture
+
+To implement the sharding strategy we need a component in charge of handling incoming alerts and API requests, distributing it to the corresponding shard. 
+
+The idea here is to have an “Alertmanager Distributor” as a first stop in the reception of alerts. Once alerts are received, the component is in charge of validating the alerts against the limits. Validated alerts are then sent to multiple managers in parallel.
+
+The individual pieces of this component (sharding, limits) cannot be optional - the optional part of it is where we decide to run it.
+
+We can either run it as a separate service or embed it. **I propose we simply embed it**. At its core it’ll be simpler to operate. With future work making it possible to run as a separate service so that operators can scale when/if needed.
+
+
+## Conclusion
```

Comment:
One thing I have not talked about is sequencing - is it worth mentioning that now? My plan is to leave the current alertmanager struct untouched and create a new one can be specified via a flag. The PRs can be slices of the sections described. Keeping in mind that I can break those down as I deemed fit to prioritise simpler and less risky reviews.

Reply:
I'd be interested in your ideas for sequencing, but seems like it could be separate from this PR.

Reply:
I'm prepping the first PR, I'll tag you in it so that you can get an idea of what my plan is.